database/connect.ts

```diff
@@ -32,7 +32,12 @@ function connectOneTimeToDatabase() {
     });
   }
 
-  return globalThis.postgresSqlClient;
+  return <T extends (object | undefined)[]>(
```

Comment:
With the last point you have here, I have been trying to rewrite the function to avoid using the generic, but i think the generic type is needed because of how we are writing the query.

Having a function like this works, but without using the generic, it throws an error `Expected 0 type arguments, but got 1.`  whenever we want to use the `sql()` function the way we use it, So we somehow need to be able to pass in an argument to the `sql()` like this `sql<Animal[]>` and i think that doing it with generic would be the only way?

```ts
  return (...sqlQuery: [TemplateStringsArray, ...any[]]) => {
    headers();
    const result = globalThis.postgresSqlClient!(...sqlQuery);
    return result as Promise<any>;
  };
```

The error it throws when the function is created without having the generic

![Screenshot 2023-06-07 at 14 13 51](https://github.com/upleveled/next-js-example-spring-2023-vienna-austria/assets/74430629/752b204f-59a8-4530-b826-71646754f40b)

Reply:
I don't totally understand what you mean, but in Postgres the sql type is defined like this

`interface Sql<TTypes extends Record<string, unknown> = {}>`

Reply:
I also found this type parameter in Postgres

```ts
(type parameter) PostgresType in <PostgresType extends Record<string, unknown>[]>(sqlQuery_0: TemplateStringsArray, ...sqlQuery_1: any[]): postgres.PendingQuery<PostgresType>
```

I don't know if this answers the question above

Reply:
https://github.com/upleveled/next-js-example-spring-2023-vienna-austria/pull/3/commits/9ac02405dcc22b7ea7ead18604e51fea63cc8493

Reply:
so as discussed offline:

1. @Josehower's `readonly (object | undefined)[]` type would be better
2. however, it only accounts one of the function types - [there is overloading on the `Sql` function](https://github.com/porsager/postgres/blob/d169712cb6a7d8e62435227fcb532d7e0ceeca06/types/index.d.ts#L641-L647):
   ```ts
    /**
     * Query helper
     * @param first Define how the helper behave
     * @param rest Other optional arguments, depending on the helper type
     * @returns An helper object usable as tagged template parameter in sql queries
     */
    <T, K extends Rest<T>>(first: T & First<T, K, TTypes[keyof TTypes]>, ...rest: K): Return<T, K>;
   ```
3. so in [2a26d87](https://github.com/upleveled/next-js-example-spring-2023-vienna-austria/pull/3/commits/2a26d87de142bee30450f2afa1cd78038f598b5b) we moved back to the casting that @Josehower originally had, but moved inside of the function